### ISSUE TITLE
Taining Time seems too long?

### ISSUE THREAD

**Issue by 91916117qq** [yanx27/JS3C-Net#17]
Hi, thank you for this brilliant work!
I want to reproduce your results, but during the training procedure, I find it would take 7 hours to train a single epoch, 100 epoch seems needing 700 hours training on a single Nvidia V100 32G gpu. My batch size is 8 , torch version is 1.6.0 and cuda version is 10.2. The remaining setting is same with the config in JS3C_default_kitti.yaml. Is there something wrong? How long do you take to train the JS3C-Net? Does the mismatch of torch and cuda version leading to some problems?
Thx

**Comment by 91916117qq** [comment]
Update: I change my env to python3.6.9, torch1.3.1 cuda 10.1 
However the training time doesn't change

**Comment by yanx27** [comment]
Hi @91916117qq , thanks for your attention.  It is strange to achieve too slow speed for you. Could you give me your operating system (Linux) or CPU configuration? In my experiment, a single Nvidia V100 32G GPU with batch size 6 only takes 3.5-4 hours/epoch, and the model will converge in at most 60 epochs, which takes about 1-1.5 weeks for training.

**Comment by 91916117qq** [comment]
Thanks for yout reply!

My operating system is Ubuntu 18.04.6 LTS (GNU/Linux 5.13.0-39-generic x86_64)
My cpu is Intel(R) Xeon(R) CPU E5-2698 v4
![image](https://user-images.githubusercontent.com/54439923/172050766-a7116f05-eb68-4db2-92fc-9f7200610078.png)

My conda env is as follows:
Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
_openmp_mutex             5.1                       1_gnu    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
bzip2                     1.0.8                h7b6447c_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
ca-certificates           2022.4.26            h06a4308_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
certifi                   2021.5.30        py36h06a4308_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
cython                    0.29.30                  pypi_0    pypi
icu                       58.2                 he6710b0_3    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
importlib-resources       5.4.0                    pypi_0    pypi
knn nanoflann             0.0.0                    pypi_0    pypi
libboost                  1.73.0              h3ff78a5_11    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
libedit                   3.1.20210910         h7f8727e_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
libffi                    3.2.1             hf484d3e_1007    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
libgcc-ng                 11.2.0               h1234567_1    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
libgomp                   11.2.0               h1234567_1    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
libstdcxx-ng              11.2.0               h1234567_1    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
lz4-c                     1.9.3                h295c915_1    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
ncurses                   6.3                  h7f8727e_2    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
numpy                     1.19.5                   pypi_0    pypi
openssl                   1.1.1o               h7f8727e_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
pg-op                     0.0.0                     dev_0    <develop>
pillow                    8.4.0                    pypi_0    pypi
pip                       21.2.2           py36h06a4308_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
python                    3.6.9                h265db76_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
pyyaml                    5.4.1                    pypi_0    pypi
readline                  7.0                  h7b6447c_5    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
scipy                     1.5.2                    pypi_0    pypi
setuptools                58.0.4           py36h06a4308_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
six                       1.16.0                   pypi_0    pypi
sparseconvnet             0.2                       dev_0    <develop>
spconv                    1.0                      pypi_0    pypi
sqlite                    3.33.0               h62c20be_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
tk                        8.6.11               h1ccaba5_1    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
torch                     1.3.1                    pypi_0    pypi
torchvision               0.4.2                    pypi_0    pypi
tqdm                      4.64.0                   pypi_0    pypi
wheel                     0.37.1             pyhd3eb1b0_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
xz                        5.2.5                h7f8727e_1    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
zipp                      3.6.0                    pypi_0    pypi
zlib                      1.2.12               h7f8727e_2    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
zstd                      1.4.9                haebb681_0    https://mirrors.bfsu.edu.cn/anaconda/pkgs/main

My cuda version is as follows:
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243

My python test is as follows:
Python 3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 19:07:31) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
True

My setup procedure is smooth except a some problem I met when installing spconv. The cmake version need to be 3.15 or higher, so I follow this blog(https://blog.csdn.net/ymzhu385/article/details/122049749?spm=1001.2014.3001.5506) to upgrade my cmake version to 3.22.4. 

I use the cloud gpu and SemanticKITTI data provided by https://www.gpushare.com/ ，you can rent a cloud gpu and test on the website, but it not free.
![image](https://user-images.githubusercontent.com/54439923/172051648-8b354ef0-aaa2-4e12-bbd3-a748a0ccee1f.png)

The data is in /hy-public/SemanticKITTI/ . I create a folder in /hy-tmp/ and excut the following command to unzip data:
cd /hy-tmp/SemanticKITTI/
unzip /hy-public/SemanticKITTI/data_odometry_calib.zip
unzip /hy-public/SemanticKITTI/data_odometry_labels.zip
unzip /hy-public/SemanticKITTI/data_odometry_velodyne.zip
unzip /hy-public/SemanticKITTI/data_odometry_voxels_all.zip

Please let me know if you need other information. Thanks again for your help.

**Comment by 91916117qq** [comment]
My train.txt is as follows:
2022-06-04 13:21:34,992 - Model - INFO - # Segmentation head parameters 2689267
2022-06-04 13:21:34,992 - Model - INFO - # Completion head parameters 509676
2022-06-04 13:22:56,181 - Model - INFO - # Segmentation head parameters 2689267
2022-06-04 13:22:56,182 - Model - INFO - # Completion head parameters 509676
2022-06-04 13:22:56,616 - Model - INFO - 
Epoch 0 (0/100):
2022-06-04 13:25:18,642 - Model - INFO - # Segmentation head parameters 2689267
2022-06-04 13:25:18,642 - Model - INFO - # Completion head parameters 509676
2022-06-04 13:25:19,071 - Model - INFO - 
Epoch 0 (0/100):
2022-06-04 20:39:56,415 - Model - INFO - Train Loss: 1.911
2022-06-04 21:27:50,248 - Model - INFO - 
  ========================== COMPLETION RESULTS ==========================  
2022-06-04 21:27:50,249 - Model - INFO - IoU class 1 [car] = 28.304
2022-06-04 21:27:50,249 - Model - INFO - IoU class 2 [bicycle] = 0.000
2022-06-04 21:27:50,250 - Model - INFO - IoU class 3 [motorcycle] = 0.000
2022-06-04 21:27:50,250 - Model - INFO - IoU class 4 [truck] = 2.095
2022-06-04 21:27:50,250 - Model - INFO - IoU class 5 [other-vehicle] = 1.653
2022-06-04 21:27:50,250 - Model - INFO - IoU class 6 [person] = 0.000
2022-06-04 21:27:50,250 - Model - INFO - IoU class 7 [bicyclist] = 0.000
2022-06-04 21:27:50,250 - Model - INFO - IoU class 8 [motorcyclist] = 0.000
2022-06-04 21:27:50,251 - Model - INFO - IoU class 9 [road] = 41.392
2022-06-04 21:27:50,251 - Model - INFO - IoU class 10 [parking] = 9.045
2022-06-04 21:27:50,251 - Model - INFO - IoU class 11 [sidewalk] = 22.177
2022-06-04 21:27:50,251 - Model - INFO - IoU class 12 [other-ground] = 0.006
2022-06-04 21:27:50,251 - Model - INFO - IoU class 13 [building] = 15.626
2022-06-04 21:27:50,251 - Model - INFO - IoU class 14 [fence] = 8.515
2022-06-04 21:27:50,251 - Model - INFO - IoU class 15 [vegetation] = 6.142
2022-06-04 21:27:50,251 - Model - INFO - IoU class 16 [trunk] = 8.264
2022-06-04 21:27:50,252 - Model - INFO - IoU class 17 [terrain] = 28.663
2022-06-04 21:27:50,252 - Model - INFO - IoU class 18 [pole] = 11.134
2022-06-04 21:27:50,252 - Model - INFO - IoU class 19 [traffic-sign] = 0.047
2022-06-04 21:27:50,252 - Model - INFO - Precision =	14.28
Recall =	94.87
IoU Cmpltn =	14.17
mIoU SSC =	9.63
2022-06-04 21:27:50,253 - Model - INFO - 
  ========================== SEGMENTATION RESULTS ==========================  
2022-06-04 21:27:50,253 - Model - INFO - IoU class 0 [car] = 86.555
2022-06-04 21:27:50,253 - Model - INFO - IoU class 1 [bicycle] = 0.000
2022-06-04 21:27:50,253 - Model - INFO - IoU class 2 [motorcycle] = 16.913
2022-06-04 21:27:50,253 - Model - INFO - IoU class 3 [truck] = 14.439
2022-06-04 21:27:50,253 - Model - INFO - IoU class 4 [other-vehicle] = 12.297
2022-06-04 21:27:50,254 - Model - INFO - IoU class 5 [person] = 17.555
2022-06-04 21:27:50,254 - Model - INFO - IoU class 6 [bicyclist] = 28.035
2022-06-04 21:27:50,254 - Model - INFO - IoU class 7 [motorcyclist] = 0.143
2022-06-04 21:27:50,254 - Model - INFO - IoU class 8 [road] = 80.340
2022-06-04 21:27:50,254 - Model - INFO - IoU class 9 [parking] = 16.051
2022-06-04 21:27:50,254 - Model - INFO - IoU class 10 [sidewalk] = 59.715
2022-06-04 21:27:50,254 - Model - INFO - IoU class 11 [other-ground] = 0.023
2022-06-04 21:27:50,254 - Model - INFO - IoU class 12 [building] = 80.342
2022-06-04 21:27:50,255 - Model - INFO - IoU class 13 [fence] = 32.378
2022-06-04 21:27:50,255 - Model - INFO - IoU class 14 [vegetation] = 83.627
2022-06-04 21:27:50,255 - Model - INFO - IoU class 15 [trunk] = 53.525
2022-06-04 21:27:50,255 - Model - INFO - IoU class 16 [terrain] = 73.071
2022-06-04 21:27:50,255 - Model - INFO - IoU class 17 [pole] = 41.787
2022-06-04 21:27:50,255 - Model - INFO - IoU class 18 [traffic-sign] = 26.215
2022-06-04 21:27:50,255 - Model - INFO - Eval point avg class IoU: 38.053294
2022-06-04 21:27:50,279 - Model - INFO - 
Best segmentation IoU: 38.053294
2022-06-04 21:27:50,279 - Model - INFO - Best semantic completion IoU: 9.634726
2022-06-04 21:27:50,279 - Model - INFO - Best completion IoU: 14.171320
2022-06-04 21:27:50,281 - Model - INFO - 
Epoch 1 (1/100):
2022-06-04 21:27:50,281 - Model - INFO - Learning rate:0.001000
2022-06-04 21:27:50,281 - Model - INFO - BN momentum updated to: 0.500000

My args.txt is as follows:
{
  "GENERAL": {
    "task": "train",
    "manual_seed": 123,
    "dataset_dir": "/hy-tmp/SemanticKITTI/dataset/",
    "debug": false
  },
  "DATA": {
    "dataset": "SemanticKITTI",
    "classes_seg": 19,
    "classes_completion": 20
  },
  "Segmentation": {
    "model_name": "SubSparseConv",
    "m": 16,
    "block_residual": false,
    "block_reps": 1,
    "seg_groups": 1,
    "use_coords": false,
    "feature_dims": [
      16,
      32,
      48,
      64,
      80,
      96,
      112
    ],
    "input_channel": 3,
    "scale": 10,
    "full_scale": [
      0,
      2048
    ],
    "max_npoint": 250000,
    "mode": 4
  },
  "Completion": {
    "model_name": "SSCNet",
    "m": 32,
    "feeding": "both",
    "no_fuse_feat": false,
    "block_residual": true,
    "block_reps": 2,
    "use_coords": false,
    "mode": 0,
    "full_scale": [
      256,
      256,
      32
    ],
    "interaction": true,
    "pooling_type": "mean",
    "fuse_k": 5,
    "point_cloud_range": [
      0,
      -25.6,
      -2,
      51.2,
      25.6,
      4.4
    ],
    "voxel_size": 0.2,
    "search_k": 8,
    "feat_relation": false
  },
  "TRAIN": {
    "epochs": 100,
    "train_workers": 10,
    "optim": "Adam",
    "batch_size": 6,
    "learning_rate": 0.001,
    "lr_decay": 0.7,
    "decay_step": 10,
    "momentum": 0.9,
    "weight_decay": 0.0001,
    "save_freq": 16,
    "uncertainty_loss": true,
    "loss_weight": [
      1.0,
      0.8
    ],
    "pretrain_path": null,
    "train_from": 0,
    "seg_num_per_class": [
      55437630,
      320797,
      541736,
      2578735,
      3274484,
      552662,
      184064,
      78858,
      240942562,
      17294618,
      170599734,
      6369672,
      230413074,
      101130274,
      476491114,
      9833174,
      129609852,
      4506626,
      1168181
    ],
    "complt_num_per_class": [
      7632350044,
      15783539,
      125136,
      118809,
      646799,
      821951,
      262978,
      283696,
      204750,
      61688703,
      4502961,
      44883650,
      2269923,
      56840218,
      15719652,
      158442623,
      2061623,
      36970522,
      1151988,
      334146
    ]
  },
  "gpu": "0",
  "config": "opt/JS3C_default_kitti.yaml",
  "log_dir": "test_first"
}

**Comment by yanx27** [comment]
Hi @91916117qq . I guess the reason is that you use too less CPU number and the process of voxelization in SparseConv highly depends on computation on CPUs. Could you try again with more CPU number and test the cost of time in each different parts (e.g., data processing, voxelization, model forward and loss)?

**Comment by yanx27** [comment]
@91916117qq Hi, did you solve the problem? It is the time to close the issue, and if you have any problem please feel free to reopen this.
